### PR TITLE
issue #3401: File.copy crashes interpreter when destination exists

### PR DESCRIPTION
**Pull request by shimpe**

### CODE REVIEW

**Comment by shimpe** [comment]
Requested changes are fine with me. Just be aware that other code in that file (prFileMkDir) uses the same code style, i.e. it uses c_str to print paths in warnings and returns errNone. Should that be changed too then?

**Comment by brianlheim** [comment]
> Just be aware that other code in that file (prFileMkDir) uses the same code style, i.e. it uses c_str to print paths in warnings and returns errNone. Should that be changed too then?

Oh, good catch! Yes, absolutely on the c_str part. Changing it to an error from a warning would be an API change but, I think, one we should make.

> Interesting to git blame those lines too (sorry couldn't resist).

haha, yeah. I was making these changes hastily and unfortunately the PR was too big to get a thorough review, so that fell through the cracks.

File: lang/LangPrimSource/PyrFilePrim.cpp [hunk 1]
```diff
@@ -195,8 +195,12 @@ int prFileCopy(struct VMGlobals * g, int numArgsPushed)
...
 	const bfs::path& p1 = SC_Codecvt::utf8_str_to_path(filename1);
 	const bfs::path& p2 = SC_Codecvt::utf8_str_to_path(filename2);
-	bfs::copy(p1, p2);
-	return errNone;
+	boost::system::error_code error_code;
+	bfs::copy(p1, p2, error_code);
+	if (error_code)
+		postfl("Warning: %s (copy from \"%s\" to \"%s\")\n", error_code.message().c_str(), filename1, filename2);
+
+	return error_code.value();
```
Comment (brianlheim) [inline lang/LangPrimSource/PyrFilePrim.cpp]: This is also a problem - error_code values don't map onto primitive return values at all - the allowed values are right at the top of the [Writing Primitives](http://doc.sccode.org/Guides/WritingPrimitives.html#Example) guide I linked earlier. The most straightforward thing to do right now (unfortunately) would be to construct the error message, then throw it in an exception. It'll be caught by the interpreter runtime, which will print the message if it's not caught.

The problem with printing like this, even if you do return a nonzero code, is that it's impossible to suppress this message even if the error is caught and handled correctly by the caller. We really need to talk about how we can improve this in the future - inside a primitive, throwing is heavy-handed and unidiomatic, but returning error codes provides very little debugging information for users.

Comment (telephon) [reply lang/LangPrimSource/PyrFilePrim.cpp]: what you can see in the primitive opcodes is that you can call an sclang message from where you want to throw the error, e.g. `primitiveFailed`. It's a special symbol in `gSpecialSelectors[primitiveFailed];`

**Pull request merged**